ibacm/src/libacm.c

```diff
@@ -59,18 +59,22 @@ static void acm_set_server_port(void)
 int ib_acm_connect(char *dest)
 {
 	struct addrinfo hint, *res;
-	int ret;
+	int ret = 0;
 
 	acm_set_server_port();
 	memset(&hint, 0, sizeof hint);
 
 	if (dest && *dest != '/') {
 		hint.ai_family = AF_INET;
-		hint.ai_family = AF_UNSPEC;
+		hint.ai_protocol = IPPROTO_TCP;
 
 		ret = getaddrinfo(dest, NULL, &hint, &res);
-		if (ret)
-			return ret;
+		if (ret) {
+			if (ret == EAI_SYSTEM)
+				return errno;
+			else
+				return -1;
```

Comment:
Why can't we return errno here?

Reply:
I don't believe that getaddrinfo() sets errno unless it returns EAI_SYSTEM. I suppose we could set errno to a valuea and return that. But personally, I think manufacturing an errno would be worse than not setting it since obviously the author(s) of getaddrinfo() had that choice themselves and opted not to set it.

Reply:
We don’t have to do anything with errno.  You are trying to define ib_acm_connect() to return an errno value.  So we could EIO or ENODATA?  I’ve not searched extensively for possible “general” errors...

OTOH  What about all the other EAI_* values?
gai_strerror() returns a string for those.  Perhaps it would be better to put the print in this function and simply return -1 or 0 to the caller?  I’ve not got the code in front of me but from your patch ib_acm_connect() is only called from 1 place right?

Reply:
You asked why we didn't return errno. I explained why. Whether we chose to set errno and return it or return an errno value seems immaterial.

As for why don't we return EIO or ENODATA - like I said, "I think manufacturing an errno would be worse than not setting one since obviously the author(s) of getaddrinfo() had that choice themselves and opted not to set it."  I don't see what that could possibly accomplish other than provide misleading information.

I had considered using gai_strerror() as you suggested, but thought I might receive negative feedback on printing errors from a function that seemed to be written expressly not to print error messages. I assume the original authors intended libacm functions to not print to stdout or stderr (though I did find an exception).

Reply:
Ah yea, you are correct WRT printing in ib_acm_connect()...

Ok the problem as I see it is that ib_acm_connect() is not following it's own "contract".  Since there is no man page for it, this becomes kind of difficult to determine.  But I think  we need to make ib_acm_connect() follow the same pattern as the library calls it calls.                    

getaddrinfo(), socket(), connect() all return some error and set errno.  Why not make all the error handling the same and follow that convention?                                                                 
             
08:14:34 > git diff
diff --git a/ibacm/src/libacm.c b/ibacm/src/libacm.c                                                                                 
index 66ed298a39ab..f79041973d54 100644                                                                                              
--- a/ibacm/src/libacm.c                                                                                                             
+++ b/ibacm/src/libacm.c
@@ -56,7 +56,7 @@ static void acm_set_server_port(void) 
 int ib_acm_connect(char *dest)
 {              
        struct addrinfo hint, *res;                                                                                                  
-       int ret;                                                                                                                     
+       int ret = 0;                                                                                                                 
 
        acm_set_server_port();                                                                                                       
        memset(&hint, 0, sizeof hint);                                                                                               
@@ -68,22 +68,18 @@ int ib_acm_connect(char *dest)                                                                                   
                                                                                                                                     
        sock = socket(res->ai_family, res->ai_socktype, res->ai_protocol);                                                           
        if (sock == -1) {
-               ret = errno;
-               goto err1; 
+               ret = -1;
+               goto freeaddr;                                                                                                       
        }       
                                                                                                                                     
        ((struct sockaddr_in *) res->ai_addr)->sin_port = htobe16(server_port);                                                      
        ret = connect(sock, res->ai_addr, res->ai_addrlen);                                                                          
-       if (ret)
-               goto err2;
-       
-       freeaddrinfo(res);                                                                                                           
-       return 0;
+       if (ret) {
+               close(sock);                                                                                                         
+               sock = -1;
+       }       
        
-err2:                                    
-       close(sock);                      
-       sock = -1;
-err1:  
+freeaddr:                                
        freeaddrinfo(res);                
        return ret;
}

Reply:
<sigh> that didnot format well.

`08:14:34 > git di
diff --git a/ibacm/src/libacm.c b/ibacm/src/libacm.c                                                                                 
index 66ed298a39ab..f79041973d54 100644                                                                                              
--- a/ibacm/src/libacm.c                                                                                                             
+++ b/ibacm/src/libacm.c
@@ -56,7 +56,7 @@ static void acm_set_server_port(void) 
 int ib_acm_connect(char *dest)
 {              
        struct addrinfo hint, *res;                                                                                                  
-       int ret;                                                                                                                     
+       int ret = 0;                                                                                                                 
 
        acm_set_server_port();                                                                                                       
        memset(&hint, 0, sizeof hint);                                                                                               
@@ -68,22 +68,18 @@ int ib_acm_connect(char *dest)                                                                                   
                                                                                                                                     
        sock = socket(res->ai_family, res->ai_socktype, res->ai_protocol);                                                           
        if (sock == -1) {
-               ret = errno;
-               goto err1; 
+               ret = -1;
+               goto freeaddr;                                                                                                       
        }       
                                                                                                                                     
        ((struct sockaddr_in *) res->ai_addr)->sin_port = htobe16(server_port);                                                      
        ret = connect(sock, res->ai_addr, res->ai_addrlen);                                                                          
-       if (ret)
-               goto err2;
-       
-       freeaddrinfo(res);                                                                                                           
-       return 0;
+       if (ret) {
+               close(sock);                                                                                                         
+               sock = -1;
+       }       
        
-err2:                                    
-       close(sock);                      
-       sock = -1;
-err1:  
+freeaddr:                                
        freeaddrinfo(res);                
        return ret;
}                                         
`

Reply:
I have no idea where that "return -1" removal line is coming from in the previous comment.  It is not part of my patch and I can't seem to edit it out of my comment.  :-(  github is not being friendly to me today...   :-(